sidecarhttp/http_api.go

```diff
@@ -189,6 +190,53 @@ func (s *SidecarApi) oneServiceHandler(response http.ResponseWriter, req *http.R
 	response.Write(jsonBytes)
 }
 
+
+
+// tombstoneServiceHandler takes service Id and mark the service as tombstone manually
+// This allow services to be marked as tombstone before actually shutting down the service.
+func (s *SidecarApi) tombstoneServiceHandler(response http.ResponseWriter, req *http.Request, params map[string]string) {
+	defer req.Body.Close()
+
+	response.Header().Set("Access-Control-Allow-Origin", "*")
+	response.Header().Set("Access-Control-Allow-Methods", "POST")
+	response.Header().Set("Content-Type", "application/json")
+
+	serviceId, ok := params["id"]
+	if !ok {
+		sendJsonError(response, 404, "Not Found - No service id provided")
+		return
+	}
+
+	if s.state == nil {
+		sendJsonError(response, 500, "Internal Server Error - Something went terribly wrong")
+		return
+	}
+
+	var instances []*service.Service
+	// Enter critical section
+	s.state.Lock()
+	defer s.state.Lock()
```

Comment:
The way to change the state is to use the slightly mis-named `AddServiceEntry()` https://github.com/Nitro/sidecar/blob/master/catalog/services_state.go#L282

That method itself will undoubtedly need to be modified to not set `ALIVE` status on services that are already in `DRAINING` status. Note that `AddServiceEntry()` is called from Gossip as well as from Discovery.

Reply:
Thinking about this a little more, I think the better way to get these events into the state is to send them into the same channel that the Memberlist delegate uses, like this: https://github.com/newrelic/sidecar/blob/3ed7ddc222794f2d136a2894eb07865901dedc0c/services_delegate.go#L53

Ideally we would add a light wrapper function to the `catalog/services_state.go` which does that, then call it from the delegate and HTTP handler instead of writing to the channel directly in the code.